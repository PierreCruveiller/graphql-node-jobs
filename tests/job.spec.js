const request = require('supertest')
const {
  migrateDatabase,
  seedDatabase,
  getNewServer,
  models,
  closeEverything,
  deleteTables,
  resetDatabase
} = require('./test-database.js')
const { checkForJobs } = require('./../lib/index')

let server = null
const uri = `http://localhost:${process.env.PORT || 8080}/graphql`

const acquireJob = variables => ({
  query: `mutation($typeList: [String!]!, $workerId: String) {
    acquireJob(
      typeList: $typeList
      workerId: $workerId
    ) {
      id
      name
      status
      output
    }
  }`,
  variables,
  operationName: null
})

const jobCreate = variables => ({
  query: `mutation($job: jobInput!) {
    jobCreate(
      job: $job
    ) {
      id
      name
      status
      output
    }
  }`,
  variables,
  operationName: null
})

/**
 * Starting the tests
 */
describe('Test the job endpoint', () => {
  beforeAll(async () => {
    await migrateDatabase()
    await seedDatabase()
    server = await getNewServer()
  })

  beforeEach(async () => {
    await resetDatabase()
  })

  afterEach(async () => {
    await deleteTables()
  })

  afterAll(async done => {
    await closeEverything(server, models, done)
  })

  it('List the jobs', async () => {
    const response = await request(server).get(
      `/graphql?query=query getJobs {
          job(order:"id") {
            id
            name
          }
        }
        &operationName=getJobs`
    )

    expect(response.body).toMatchSnapshot()
  })

  it('Acquiring a job require a type', async () => {
    const response = await request(server)
      .post('/graphql')
      .send(acquireJob({}))

    expect(response.body.errors).toHaveLength(1)
    expect(response.body.errors[0].message).toMatchSnapshot()
  })

  it('One can acquire a job of a given type.', async () => {
    const response = await request(server)
      .post('/graphql')
      .send(
        acquireJob({
          typeList: ['a']
        })
      )

    expect(response.body.errors).toBeUndefined()
    expect(response.body.data).toMatchSnapshot()

    const response2 = await request(server)
      .post('/graphql')
      .send(
        acquireJob({
          typeList: ['a']
        })
      )

    expect(response2.body.errors).toBeUndefined()
    expect(response2.body.data.acquireJob).toBe(null)
  })

  it('checkForJobs allows to simply acquire and update jobs.', async () => {
    const job = await checkForJobs({
      typeList: ['a'],
      uri,
      processingFunction: job => {
        return { total: 125 }
      },
      looping: false
    })
    expect(job).not.toBeUndefined()
    expect(job).not.toBe(null)
    expect(job).toMatchSnapshot()
  })

  it('checkForJobs allows to simply acquire multiple types of jobs at the same time.', async () => {
    const job = await checkForJobs({
      typeList: ['a', 'b'],
      uri,
      processingFunction: job => {
        return { total: 125 }
      },
      looping: false
    })
    expect(job).not.toBeUndefined()
    expect(job).not.toBe(null)
    expect(job).toMatchSnapshot()

    const job2 = await checkForJobs({
      typeList: ['a', 'b'],
      uri,
      processingFunction: job => {
        return { total: 125 }
      },
      looping: false
    })
    expect(job2).not.toBeUndefined()
    expect(job2).not.toBe(null)
    expect(job2).toMatchSnapshot()
  })

  it('checkForJobs allows to asynchronous processing functions.', async () => {
    const job = await checkForJobs({
      typeList: ['a', 'b'],
      uri,
      processingFunction: async job => {
        const result = await new Promise((resolve, reject) =>
          setTimeout(() => {
            resolve('plop')
          }, 100)
        )
        return { total: result }
      },
      looping: false
    })
    expect(job).not.toBeUndefined()
    expect(job).not.toBe(null)
    expect(job).toMatchSnapshot()

    const jobEntity = await models.job.findOne({ where: { id: job.id } })
    expect(jobEntity.startedAt).not.toBe(null)
    expect(jobEntity.endedAt).not.toBe(null)
  })

  it('One can create a job of a given type.', async () => {
    const response = await request(server)
      .post('/graphql')
      .send(
        jobCreate({
          job: { name: 'c', type: 'c' }
        })
      )

    expect(response.body.errors).toBeUndefined()
    expect(response.body.data).toMatchSnapshot()
    // By default a created job is always in queued state.
    expect(response.body.data.jobCreate.status).toBe('queued')
  })

  it('One cannot create a job without a type.', async () => {
    const response = await request(server)
      .post('/graphql')
      .send(
        jobCreate({
          job: { name: 'c' }
        })
      )

    expect(response.body.errors).not.toBeUndefined()
  })

  it('One cannot create a job without a type.', async () => {
    const response = await request(server)
      .post('/graphql')
      .send(
        jobCreate({
          job: { name: 'c' }
        })
      )

    expect(response.body.errors).not.toBeUndefined()
  })

  it('One can query the timefields', async () => {
    const response = await request(server).get(
      `/graphql?query=query getJobs {
          job(order:"id") {
            id
            startedAt
            endedAt
            createdAt
            deletedAt
          }
        }
        &operationName=getJobs`
    )

    expect(response.body.errors).toBeUndefined()
  })
})
