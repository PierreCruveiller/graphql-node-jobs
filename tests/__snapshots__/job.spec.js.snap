// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test the job endpoint Acquiring a job require a type 1`] = `"Variable \\"$typeList\\" of required type \\"[String!]!\\" was not provided."`;

exports[`Test the job endpoint List the jobs 1`] = `
Object {
  "data": Object {
    "job": Array [
      Object {
        "id": 1,
        "name": "Bla a",
      },
      Object {
        "id": 2,
        "name": "Bla b",
      },
    ],
  },
}
`;

exports[`Test the job endpoint One can acquire a job of a given type. 1`] = `
Object {
  "acquireJob": Object {
    "id": 1,
    "name": "Bla a",
    "output": Object {},
    "status": "processing",
  },
}
`;

exports[`Test the job endpoint One can create a job of a given type. 1`] = `
Object {
  "jobCreate": Object {
    "id": 3,
    "name": "c",
    "output": null,
    "status": "queued",
  },
}
`;

exports[`Test the job endpoint The processingFunction expose . 1`] = `
Object {
  "__typename": "job",
  "id": 1,
  "input": Object {},
  "name": "Bla a",
  "output": Object {},
  "processingInfo": Object {
    "percent": 10,
  },
  "status": "successful",
  "type": "a",
}
`;

exports[`Test the job endpoint The processingFunction expose . 2`] = `Object {}`;

exports[`Test the job endpoint The processingFunction expose . 3`] = `
Object {
  "percent": 10,
}
`;

exports[`Test the job endpoint checkForJobs allows to asynchronous processing functions. 1`] = `
Object {
  "__typename": "job",
  "id": 1,
  "input": Object {},
  "name": "Bla a",
  "output": Object {
    "total": "plop",
  },
  "processingInfo": null,
  "status": "successful",
  "type": "a",
}
`;

exports[`Test the job endpoint checkForJobs allows to simply acquire and update jobs. 1`] = `
Object {
  "__typename": "job",
  "id": 1,
  "input": Object {},
  "name": "Bla a",
  "output": Object {
    "total": 125,
  },
  "processingInfo": null,
  "status": "successful",
  "type": "a",
}
`;

exports[`Test the job endpoint checkForJobs allows to simply acquire multiple types of jobs at the same time. 1`] = `
Object {
  "__typename": "job",
  "id": 1,
  "input": Object {},
  "name": "Bla a",
  "output": Object {
    "total": 125,
  },
  "processingInfo": null,
  "status": "successful",
  "type": "a",
}
`;

exports[`Test the job endpoint checkForJobs allows to simply acquire multiple types of jobs at the same time. 2`] = `
Object {
  "__typename": "job",
  "id": 2,
  "input": Object {},
  "name": "Bla b",
  "output": Object {
    "total": 125,
  },
  "processingInfo": null,
  "status": "successful",
  "type": "b",
}
`;

exports[`Test the job endpoint checkForJobs declare jobs as failed when an error is raised. 1`] = `[TypeError: a.awd is not a function]`;

exports[`Test the job endpoint checkForJobs processing function can return nothing if needed. 1`] = `
Object {
  "__typename": "job",
  "id": 1,
  "input": Object {},
  "name": "Bla a",
  "output": Object {},
  "processingInfo": null,
  "status": "successful",
  "type": "a",
}
`;

exports[`Test the job endpoint checkForJobs processing function can return nothing if needed. 2`] = `Object {}`;
